File: CourseraAlgorithmicToolbox Resources/week4_divide_and_conquer/4_number_of_inversions/by_learners/inversions.rb
#!/usr/bin/env ruby
# by Andronik Ordian

# helper function for faster copy
# b[l..r - 1] = a[l..r]
def copy(a, b, l, r)
  while l < r
    b[l] = a[l]
    l += 1
  end
end

def get_number_of_inversions(a, b, left, right)
  number_of_inversions = 0
  return 0 if right - left <= 1

  mid = left + (right - left) / 2
  number_of_inversions += get_number_of_inversions(a, b, left, mid)
  number_of_inversions += get_number_of_inversions(a, b, mid, right)
  #write your code here

  return number_of_inversions
end

if __FILE__ == $0
  n, *a = STDIN.read.split().map(&:to_i)
  b = Array.new(n, 0)
  puts get_number_of_inversions(a, b, 0, n)
end

File: Resources/CourseraAlgorithmicToolbox/week4_divide_and_conquer/4_number_of_inversions/Inversions.kt
import java.util.*

fun getNumberOfInversions(a: IntArray, b: IntArray, left: Int, right: Int): Long {
    var numberOfInversions: Long = 0
    if (right <= left + 1) {
        return numberOfInversions
    }
    val ave = (left + right) / 2
    numberOfInversions += getNumberOfInversions(a, b, left, ave)
    numberOfInversions += getNumberOfInversions(a, b, ave, right)
    //write your code here
    return numberOfInversions
}

fun main(args: Array<String>) {
    val scanner = Scanner(System.`in`)
    val n = scanner.nextInt()
    val a = IntArray(n)
    for (i in 0 until n) {
        a[i] = scanner.nextInt()
    }
    val b = IntArray(n)
    println(getNumberOfInversions(a, b, 0, a.size))
}